### PR TITLE
[Baseline] Introduce Link prediction lightGBM baseline

### PR DESCRIPTION
**Pull request by XinweiHe**
Summary
Introduce lightGBM model for link prediction problems. During training and validation, for each known lhs and rhs entity link, we include a negative link through negative sampling. We transform the link prediction problem into binary classification problem predicting whether there is a link between each lhs, rhs entity pairs. During testing we generate the candidate set through past links and global popularity with size no larger than `2 * eval_k`, then using the lightGBM model to predict whether there is a link between each lhs, rhs entity pairs, next with the prediction score we generate a ranked list of rhs entities for each lhs entity and evaluate.

Testing
`python examples/lightgbm_link_baseline.py --dataset rel-stackex --task rel-stackex-comment-on-post`
```
[I 2024-06-02 01:42:44,338] Trial 9 finished with value: 0.5777116285006785 and parameters: {'max_depth': 10, 'learning_rate': 0.06232235167007196, 'num_leaves': 906, 'subsample': 0.14482239176013254, 'colsample_bytree': 0.9006872199897339, 'lambda_l1': 0.011682283412980651, 'lambda_l2': 2.7464931135425744e-06, 'min_data_in_leaf': 61}. Best is trial 0 with value: 0.6305251110188359.
Train: {'link_prediction_precision': 0.020272140872922458, 'link_prediction_recall': 0.9998930818786586, 'link_prediction_map': 0.9734307172752725}
Val: {'link_prediction_precision': 0.019006060606060606, 'link_prediction_recall': 0.9995191301812694, 'link_prediction_map': 0.7809986327978464}
Test: {'link_prediction_precision': 0.0002902374670184697, 'link_prediction_recall': 0.009929680490754524, 'link_prediction_map': 0.00042037854481603816}
(venv) xinwei@ip-10-0-150-242:~/code/relbench$ python examples/lightgbm_link_baseline.py
```

`python examples/lightgbm_link_baseline.py --dataset rel-trial --task rel-trial-sponsor-condition`
```
[I 2024-06-02 01:56:31,210] Trial 9 finished with value: 0.7496392223475724 and parameters: {'max_depth': 4, 'learning_rate': 0.007636588753818599, 'num_leaves': 678, 'subsample': 0.5865728712336551, 'colsample_bytree': 0.140933846101759, 'lambda_l1': 2.213740877988149e-06, 'lambda_l2': 0.0019725336286806655, 'min_data_in_leaf': 70}. Best is trial 1 with value: 0.8191616136797515.
[2000]  valid_0's auc: 0.819162
Train: {'link_prediction_precision': 0.5010694752802297, 'link_prediction_recall': 0.8446192962101553, 'link_prediction_map': 0.9689995459161604}
Val: {'link_prediction_precision': 0.46391158097068713, 'link_prediction_recall': 0.8266646106137138, 'link_prediction_map': 0.9041867410774443}
Test: {'link_prediction_precision': 0.03684978123480798, 'link_prediction_recall': 0.06498032334799916, 'link_prediction_map': 0.031915148137614396}
(venv) xinwei@ip-10-0-150-242:~/code/relbench$ python examples/lightgbm_link_baseline.py --dataset rel-trial --task rel-trial-sponsor-condition
```

### CODE REVIEW

File: examples/idgnn_link.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def test(loader: NeighborLoader) -> np.ndarray:
             f"Val metrics: {val_metrics}"
         )
 
-        if val_metrics[tune_metric] > best_val_metric:
+        if val_metrics[tune_metric] >= best_val_metric:
```
Comment (weihua916) [inline examples/idgnn_link.py:188]: Don't change

File: examples/lightgbm_link_baseline.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+# Prepare col_to_stype dictioanry mapping between column names and stypes
+# for torch_frame Dataset initialization.
+col_to_stype = {}
+lhs_entity_table_col_to_stype = dataset2inferred_stypes[args.dataset][
+    task.src_entity_table
+]
+rhs_entity_table_col_to_stype = dataset2inferred_stypes[args.dataset][
+    task.dst_entity_table
+]
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:55]: Don't use `lhs`, `rhs` terminology.
Use `src`, `dst` instead.

**Comment by weihua916** [comment]
This is a very cool and clever baseline!

File: examples/lightgbm_link_baseline.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+)
+for column_name in lhs_rhs_intersection_column_names:
+    lhs_entity_table_col_to_stype[f"{column_name}_x"] = lhs_entity_table_col_to_stype[
+        column_name
+    ]
+    del lhs_entity_table_col_to_stype[column_name]
+    rhs_entity_table_col_to_stype[f"{column_name}_y"] = rhs_entity_table_col_to_stype[
+        column_name
+    ]
+    del rhs_entity_table_col_to_stype[column_name]
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:77]: Can you write some comments why you do this? If we rename this, don't we need to change the column name of the original data frame as well?

File: examples/lightgbm_link_baseline.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+]
+
+if lhs_entity_table.pkey_col is not None:
+    del lhs_entity_table_col_to_stype[lhs_entity_table.pkey_col]
+for fkey_col in lhs_entity_table.fkey_col_to_pkey_table.keys():
+    del lhs_entity_table_col_to_stype[fkey_col]
+if rhs_entity_table.pkey_col is not None:
+    del rhs_entity_table_col_to_stype[rhs_entity_table.pkey_col]
+for fkey_col in rhs_entity_table.fkey_col_to_pkey_table.keys():
+    del rhs_entity_table_col_to_stype[fkey_col]
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:64]: There is a similar logic [here](https://github.com/snap-stanford/relbench/blob/main/relbench/external/graph.py#L79-L86).

Can we create some util function [here](https://github.com/snap-stanford/relbench/blob/main/relbench/external/utils.py)? something like 
```python
def remove_pkey_fkey(col_to_stype, table) -> dict:
   # return col_to_stype where pkey and fkey are removed.
```

File: examples/lightgbm_link_baseline.py (hunk 3 above)
Comment (XinweiHe) [reply examples/lightgbm_link_baseline.py:77]: Sure, just added some comments for this.

File: examples/lightgbm_link_baseline.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+grouped_deduped_trainval = (
+    trainval_table_df.groupby(left_entity)[right_entity]
+    .apply(dst_entities_aggr)
+    .reset_index()
+)
+test_df = pd.merge(test_df, grouped_deduped_trainval, how="left", on=left_entity)
+
+# collect the most popular rhs entities
+all_rhs_entities = [
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:160]: still a bunch of rhs/lhs terminologies to remove

**Review by weihua916 — APPROVED** [review]
Thanks. Approving to unblock. It'd be nice for @rishabh-ranjan or @rusty1s to have a second look.

**Comment by weihua916** [comment]
Personal opinion, I feel only generating  `2 * eval_k` candidates is quite restrictive. Can we generate `10 * eval_k` instead?

File: examples/lightgbm_link_baseline.py (hunk 5 above)
Comment (XinweiHe) [reply examples/lightgbm_link_baseline.py:160]: Fixing it right now

**Comment by XinweiHe** [comment]
> Personal opinion, I feel only generating 2 * eval_k candidates is quite restrictive. Can we generate 10 * eval_k instead?

done

**Comment by XinweiHe** [comment]
Tested `10 * eval_k` with `python examples/lightgbm_link_baseline.py --dataset rel-stackex --task rel-stackex-comment-on-post` and got
Test: {'link_prediction_precision': 9.234828496042217e-05, 'link_prediction_recall': 0.0034613195795874666, 'link_prediction_map': 0.000157805313375241}
worse results compared to `2 * eval_k`. Let's maybe keep `2 * eval_k` for now.

**Pull request merged**